README.md
# DDC

DDC (Data Driven Controllers) lets you declare how to wire Rails into your app without the need for code. A Rails controller's job is parsing/interpreting parameters to send to your application domain and taking those results and translating them back out to an HTTP result (html/status/headers). DDC removes the need for all the boiler plate controller code and tests.

By adhering to a couple of interfaces, you can avoid writing most controller code and tests. See this [blog post]( http://spin.atomicobject.com/2015/01/26/data-driven-rails-controllers) for more information.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'ddc'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install ddc

## Usage

### Controllers

`controllers/monkeys_controller.rb`

```ruby
DDC::ControllerBuilder.build :monkeys,
  before_actions: [:authenticate_user!],
  actions: {
    show: {
      context: 'context_builder#user_and_id',
      service: 'monkey_service#find'
    },
    index: {
      context: 'context_builder#user',
      service: 'monkey_service#find_all'
    },
    update: {
      context: 'context_builder#monkey',
      service: 'monkey_service#update'
    },
    create: {
      context: 'context_builder#monkey',
      service: 'monkey_service#create'
    }
  }
```

### Context Builders
`lib/context_builder.rb`

```ruby
class ContextBuilder
  def user(context_params)
    HashWithIndifferentAccess.new current_user: context_params[:current_user] 
  end

  def user_and_id(context_params)
    user(context_params).merge(id: context_params[:params][:id])
  end

  def monkey(context_params)
    info = context_params[:params].permit(monkey: [:color, :poo])
    user_and_id(context_params).merge(info)
  end
end
```


### Services

`lib/monkeys_service.rb`

```ruby
class MonkeyService
  def find(context)
    id, user = context.values_at :id, :current_user
    me = find_for_user user, id
    if me.present?
      ok(me)
    else
      not_found
    end
  end

  def update(context)
    id, user, updates = context.values_at :id, :current_user, @model_type
    me = find_for_user user, id

    translated_updates = translated_cid_to_id(updates)
    
    if me.present?
      me.update_attributes translated_updates
      ok(me)
    else
      not_found
    end
  end

  private
  def not_found
    {status: :not_found}.freeze
  end
  def ok(obj)
    {status: :ok, object: obj}
  end
end

# shortcut for default CRUD service
MonkeyService = DDC::ServiceBuilder.build(:monkey)
```

## Contributing

1. Fork it ( https://github.com/atomicobject/ddc/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# DDC Changelog

### 0.1.9

* Rails 5 compat

### 0.1.8

* ResponseBuilder#delete no longer requires arg

### 0.1.7

* patch to fix 0.1.6 bug

### 0.1.6

* remove use of `qualified_const_*` calls to fix a bug with global lookup
* add :parent option

### 0.1.5

* remove :render_opts to use :object_render_opts
* add :errors_render_opts

### 0.1.4

* remove :serializer and :each_serializer,  use :render_opts to pass a hash

### 0.1.2

* add :serializer and :each_serializer as action options

### 0.1.1

* Fix update status code
* Add missing delete for generated services

### 0.1.0

* API updates to be more consistent
* Add ServiceBuilder that delegats to AR Model or XxxxxFinder

### 0.0.1

* Initial release


spec/spec_helper.rb
require 'active_support/all'
require 'action_controller'

class ApplicationController < ActionController::Base
end

require 'pry'
require_relative '../lib/ddc'



spec/service_builder_spec.rb
require 'spec_helper'

describe DDC::ServiceBuilder do
  subject { described_class }

  describe '.build' do
    before do
      Object.send(:remove_const, :FooFinder) if defined? FooFinder
    end

    class Foo
    end

    it 'defines a class that responds to CRUD' do
      foo_service_klass = subject.build :foo
      service = foo_service_klass.new
      expect(service).to respond_to(:find)
      expect(service).to respond_to(:find_all)
      expect(service).to respond_to(:update)
      expect(service).to respond_to(:create)
    end

    describe '#find' do
      it 'uses the AR model by default and returns DDC status hash' do
        foo_service_klass = subject.build :foo
        service = foo_service_klass.new

        expect(Foo).to receive(:where).with(id: 'monkey').and_return :myfoo
        found_data = service.find({id: 'monkey'})
        expect(found_data[:object]).to eq(:myfoo)
        expect(found_data[:status]).to eq(:ok)
      end

      it 'returns a not_found status' do
        foo_service_klass = subject.build :foo
        service = foo_service_klass.new

        expect(Foo).to receive(:where).with(id: 'monkey').and_return nil
        found_data = service.find({id: 'monkey'})
        expect(found_data[:object]).to eq(nil)
        expect(found_data[:status]).to eq(:not_found)
      end

      it 'uses the FooFinder if present' do
        FooFinder = Class.new
        foo_service_klass = subject.build :foo
        service = foo_service_klass.new

        expect(FooFinder).to receive(:find).with(id: 'monkey').and_return :myfoo
        found_data = service.find({id: 'monkey'})
        expect(found_data[:object]).to eq(:myfoo)
        expect(found_data[:status]).to eq(:ok)
      end
    end

    describe '#find_all' do
      it 'uses the AR model by default and returns DDC status hash' do
        foo_service_klass = subject.build :foo
        service = foo_service_klass.new

        expect(Foo).to receive(:all).and_return :myfoo
        found_data = service.find_all
        expect(found_data[:object]).to eq(:myfoo)
        expect(found_data[:status]).to eq(:ok)
      end

      it 'uses the FooFinder if present' do
        FooFinder = Class.new
        foo_service_klass = subject.build :foo
        service = foo_service_klass.new

        expect(FooFinder).to receive(:find_all).with(user: 'monkey').and_return :myfoo
        found_data = service.find_all({user: 'monkey'})
        expect(found_data[:object]).to eq(:myfoo)
        expect(found_data[:status]).to eq(:ok)
      end
    end
  end
end



spec/controller_builder_spec.rb
require 'spec_helper'

describe DDC::ControllerBuilder do
  subject { described_class }

  describe '.build' do
    let(:json_format) {
      format = double('json format')
      expect(format).to receive(:json).and_yield
      expect(format).to receive(:html)
      format }
    let(:html_format) {
      format = double('html format')
      expect(format).to receive(:html).and_yield
      expect(format).to receive(:json)
      format }
    after do
      klass_to_cleanup = :FooController
      Object.send :remove_const, klass_to_cleanup if Object.constants.include?(klass_to_cleanup)
    end

    it 'defines the controller class' do
      subject.build :foo, actions: {
        index: {
          context_params: [:current_user, :params],
          context: 'foo_context_builder#bar',
          service: 'baz_service#qux'
        }
      }
      expect(Object.const_get("FooController")).not_to be_nil
    end

    it 'raises if there are no actions defined' do
      expect(->{subject.build :foo, actions: {}}).to raise_exception(RuntimeError, "Must specify actions")
      expect(->{subject.build :foop, {}}).to raise_exception(RuntimeError, "Must specify actions")
    end

    it 'raises if an action is missing context' do
      expect(->{subject.build :foo, actions: {foo: {
        context_params: [:current_user, :params],
        service: 'baz_service#qux'
      }}}).to raise_exception(RuntimeError, "Must specify a context for each action")
    end

    it 'raises if an action is missing service' do
      expect(->{subject.build :foo, actions: {foo: {
        context: 'foo_context_builder#bar',
      }}}).to raise_exception(RuntimeError, "Must specify a service for each action")
    end

    it 'uses the provided parent class' do
      class DarthVader
      end

      klass = subject.build :foo,
        parent: DarthVader,
        actions: {
          index: {
            context: 'foo_context_builder#bar',
            service: 'baz_service#qux'
          }
        }
      expect(klass.ancestors[1]).to be(DarthVader)
    end

    it 'adds the before actions' do
      class FooController
        def self.before_action(*args);end
      end

      expect(FooController).to receive(:before_action).with(:my_before_action)
      subject.build :foo,
        before_actions: [:my_before_action],
        actions: {
          index: {
            context: 'foo_context_builder#bar',
            service: 'baz_service#qux'
          }
        }

    end

    it 'sunny day get params, process, return object and status, render' do
      class FooController
        def current_user; end
        def some_user; end
        def render(args); end
        def respond_to; end
      end
      controller = FooController.new

      expect(controller).to receive_messages(
        current_user: :some_user,
        params: {a: :b})

      render_args = nil
      expect(controller).to receive(:render) do |args|
        render_args = args
      end
      expect(controller).to receive(:respond_to) do |&block|
        block.call(json_format)
      end
      expect_any_instance_of(FooContextBuilder).to receive(:bar).with(hash_including(
        current_user: :some_user,
        params: {a: :b})) { :context }

      expect_any_instance_of(BazService).to receive(:qux).with(:context) do
        { object: :some_obj, status: :ok }
      end

      subject.build :foo, actions: {
        index: {
          context_params: [:current_user, :params],
          context: 'foo_context_builder#bar',
          service: 'baz_service#qux'
        }
      }
      controller.index

      expect(render_args).to eq(json: :some_obj, status: 200)
    end

    it 'renders error if service returns nil object' do
      class FooController
        def current_user; end
        def some_user; end
        def render(args); end
        def respond_to; end
      end

      subject.build :foo, actions: {
        index: {
          context_params: [:current_user, :params],
          context: 'foo_context_builder#bar',
          service: 'baz_service#qux'
        }
      }
      controller = FooController.new
      expect(controller).to receive_messages(
        current_user: :some_user,
        params: {a: :b})

      render_args = nil
      expect(controller).to receive(:render) do |args|
        render_args = args
      end
      expect(controller).to receive(:respond_to) do |&block|
        block.call(json_format)
      end

      expect_any_instance_of(FooContextBuilder).to receive(:bar).with(hash_including(
        current_user: :some_user,
        params: {a: :b})) { :context }

      expect_any_instance_of(BazService).to receive(:qux).with(:context) do
        { status: :error, errors: ["BOOM"] }
      end

      controller.index
      expect(render_args).to eq(json: {errors: ["BOOM"]}, status: 500)
    end

    it 'defines all the action methods' do
      class FooController
        def current_user; end
        def some_user; end
        def render(args); end
      end
      subject.build :foo,
        params: [:current_user, :params],
        actions: {
          index: {
            context: 'foo_context_builder#bar',
            service: 'baz_service#qux'
          },
          other: {
            context: 'foo_context_builder#bar',
            service: 'baz_service#qux'
          }
        }
      controller = FooController.new
      expect(controller).to respond_to(:index)
      expect(controller).to respond_to(:other)
    end

    class FooContextBuilder
      def bar(opts) {} end
    end

    class BazService
      def qux(context) {} end
    end

    context 'render_opts' do

      let (:controller) {
        class FooController
          def current_user; end
          def some_user; end
          def render(args); end
          def respond_to; end
        end
        FooController.new
      }
      before do
        expect(controller).to receive_messages( params: {a: :b})

        subject.build :foo,
          actions: {
            index: {
              context: 'foo_context_builder#bar',
              service: 'baz_service#qux',
              render_opts: {
                serializer: MySerializer
              }
            },
            show: {
              context: 'foo_context_builder#bar',
              service: 'baz_service#qux',
              render_opts: {
                serializer: MySerializer
              },
              object_render_opts: {
                serializer: MyObjectSerializer
              },
              error_render_opts: {
                serializer: MyErrorSerializer
              },
            }

          }

        @render_args = nil
        expect(controller).to receive(:render) do |args|
          @render_args = args
        end
        expect(controller).to receive(:respond_to) do |&block|
          block.call(json_format)
        end
        expect_any_instance_of(FooContextBuilder).to receive(:bar).with(hash_including(
          params: {a: :b})) { :context }
      end

      it 'uses specified serializer json render calls' do
        expect_any_instance_of(BazService).to receive(:qux).with(:context) do
          { object: :some_obj, status: :ok }
        end
        controller.index
        expect(@render_args[:serializer]).to eq(MySerializer)
      end

      it 'uses object_render_opts' do
        expect_any_instance_of(BazService).to receive(:qux).with(:context) do
          { object: :some_obj, status: :ok }
        end
        controller.show
        expect(@render_args[:serializer]).to eq(MyObjectSerializer)
      end

      it 'uses error_render_opts' do
        expect_any_instance_of(BazService).to receive(:qux).with(:context) do
          { errors: [], status: :not_valid }
        end
        controller.show
        expect(@render_args[:serializer]).to eq(MyErrorSerializer)
      end
    end

    class MySerializer
    end

    class MyObjectSerializer
    end

    class MyErrorSerializer
    end

    class FooContextBuilder
      def bar(opts) {} end
    end

    class BazService
      def qux(context) {} end
    end

    class MultiContextBuilder
      def foo(opts) {foo: 2} end
      def bar(opts) {bar: 3} end
    end

    class MultiContextService
      def check(ctx)
        {status: :ok, object: ctx[:foo] + ctx[:bar]}
      end
    end

    it 'supports multiple contexts' do
      class FooController
        def render(args); end
        def respond_to; end
      end
      subject.build :foo, actions: {
        index: {
          context_params: [],
          contexts: ['multi_context_builder#foo', 'multi_context_builder#bar'],
          service: 'multi_context_service#check'
        }
      }
      expect(Object.const_get("FooController")).not_to be_nil
      controller = Object.const_get("FooController").new
      render_args = nil
      expect(controller).to receive(:render) do |args|
        render_args = args
      end
      expect(controller).to receive(:respond_to) do |&block|
        block.call(json_format)
      end
      controller.index
      expect(render_args[:json]).to eq(5)
    end
  end
end



lib/ddc/service_builder.rb
module DDC
  class ServiceBuilder
    def self.build(model_type)
      Class.new do
        include DDC::ResponseBuilder
        class << self
          attr_accessor :model_type, :ar_model, :finder
        end

        @model_type = model_type
        ar_class_name = model_type.to_s.camelize
        @ar_model = Object.const_get(ar_class_name)

        @finder = nil
        begin
          @finder = Object.const_get("#{ar_class_name}Finder")
        rescue NameError
          # we use the AR Model as a fallback 
        end

        def find(context)
          id = context[:id]
          me = custom_finder ? custom_finder.find(context) : 
                               ar_model.where(id: id)
          if me.present?
            ok(me)
          else
            not_found
          end
        end

        def find_all(context={})
          mes = custom_finder ? custom_finder.find_all(context) : 
                                ar_model.all
          ok(mes)
        end

        def update(context)
          id, updates = context.values_at :id, self.class.model_type
          me = self.class.ar_model.where id: id

          if me.present?
            success = me.update_attributes translated_updates
            if success
              ok(me)
            else
              not_valid(me)
            end
          else
            not_found
          end
        end

        def create(context)
          attributes = context.values_at self.class.model_type
          me = self.class.ar_model.create attributes
          if me.persisted?
            created(me)
          else
            not_valid(me)
          end
        end

        def delete(context)
          id = context[:id]
          me = custom_finder ? custom_finder.find(context) : 
                               ar_model.where(id: id)
          if me.present?
            me.destroy
            deleted
          else
            not_found
          end
        end

        private
        def custom_finder
          self.class.finder
        end

        def ar_model
          self.class.ar_model
        end

      end
    end
  end
end


lib/ddc/controller_builder.rb
module DDC
  class ControllerBuilder
    DEFAULT_CONTEXT_PARAMS = [:params]
    DEFAULT_STATUSES = {
      ok: 200,
      created: 201,
      not_found: 404,
      not_allowed: 401,
      error: 500,
      not_valid: 422,
      deleted: 204
    }
    class << self
      def build(controller_name, config)
        klass = find_or_create_class(controller_name, config)
        setup_before_actions!(klass, config)
        setup_actions!(controller_name, klass, config)
        klass
      end

 
      def specific_const_defined?(path)
        path.split("::").inject(Object) do |mod, name|
          return unless mod.const_defined?(name, false)
          mod.const_get(name, false)
        end
        return true
      end

      def specific_const_get(path)
        path.split("::").inject(Object) do |mod, name|
          mod.const_get(name, false)
        end
      end

      def specific_const_set(path, klass)
        path_pieces = path.split("::")
        mod_name = path_pieces[0..-2].join("::")
        mod = mod_name.present? ? Object.const_get(mod_name) : Object
        mod.const_set(path_pieces.last, klass)
      end

      def find_or_create_class(controller_name, config)
        controller_klass_name = controller_name.to_s.camelize+'Controller'
        klass = nil
        if specific_const_defined?(controller_klass_name)
          klass = specific_const_get(controller_klass_name)
        else
          parent_klass = config[:parent] || ApplicationController
          klass = Class.new(parent_klass)
          specific_const_set(controller_klass_name, klass)
        end
        klass
      end
 
      def setup_before_actions!(klass, config)
        (config[:before_actions] || []).each do |ba|
          klass.before_action ba
        end
      end

      def setup_actions!(controller_name, klass, config)
        actions = config[:actions]
        raise "Must specify actions" if actions.blank?

        actions.each do |action, action_desc|
          setup_action! controller_name, klass, action, action_desc, config
        end
      end

      def reduce_contexts(contexts, context_params)
        computed_contexts = contexts.map do |context_klass, context_method|
          context_klass.new.send(context_method, context_params)
        end
        if computed_contexts.size == 1
          computed_contexts.first
        else
          computed_contexts.reduce({}) do |h, ctx|
            h.merge(ctx)
          end.with_indifferent_access
        end
      end

      def setup_action!(controller_name, klass, action, action_desc, config)
        raise "Must specify a service for each action" unless action_desc[:service].present?
        raise "Must specify a context for each action" unless (action_desc[:context].present? || action_desc[:contexts].present?)
        proc_klass, proc_method = parse_class_and_method(action_desc[:service])
        contexts = (action_desc[:contexts] || [action_desc[:context]]).map { |ctx| parse_class_and_method ctx }
        #context_klass, context_method = parse_class_and_method(action_desc[:context])

        klass.send :define_method, action do
          context_params = (action_desc[:context_params] || config[:context_params] || DEFAULT_CONTEXT_PARAMS).inject({}) do |h, param|
            h[param] = send param
            h
          end
          context = DDC::ControllerBuilder.reduce_contexts contexts, context_params

          result = proc_klass.new.send(proc_method, context)
          obj = result[:object]
          errors = result[:errors] || []
          plural_model_name = controller_name.to_s
          model_name = plural_model_name.singularize

          # alias in object as model name
          if obj.is_a? Enumerable
            result[plural_model_name] ||= obj
          else
            result[model_name] ||= obj
          end

          status = DEFAULT_STATUSES.merge(action_desc[:status]||{})[result[:status]]

          respond_to do |format|
            format.json do
              if obj.nil?
                render_opts = { json: {errors: errors}, status: status }
                render_opts.reverse_merge!(action_desc[:error_render_opts]) if action_desc.has_key? :error_render_opts
              else
                render_opts = { json: obj, status: status }
                render_opts.reverse_merge!(action_desc[:object_render_opts]) if action_desc.has_key? :object_render_opts
              end

              render_opts.reverse_merge!(action_desc[:render_opts]) if action_desc.has_key? :render_opts
              render render_opts
            end
            format.html do
              result.each do |k,v|
                instance_variable_set("@#{k}", v)
              end
            end
          end
        end
      end

      def parse_class_and_method(str)
        under_klass, method = str.split('#')
        [Object.const_get(under_klass.camelize), method]
      end
    end
  end

end
